File: src/db/fx.rs
use sqlx::SqlitePool;
use chrono::NaiveDate;

pub async fn get_latest_usdjpy(pool: &SqlitePool) -> Result<Option<f64>, sqlx::Error> {
    let row: Option<(f64,)> = sqlx::query_as(
        r#"
        SELECT rate FROM fx_cache WHERE pair = 'USDJPY' ORDER BY date DESC LIMIT 1
        "#
    )
    .fetch_optional(pool)
    .await?;

    Ok(row.map(|r| r.0))
}

pub async fn bulk_insert_usdjpy(
    pool: &SqlitePool,
    rows: &[(NaiveDate, f64)],
) -> Result<usize, sqlx::Error> {
    if rows.is_empty() {
        return Ok(0);
    }
    let mut tx = pool.begin().await?;
    let mut inserted = 0usize;
    for (date, rate) in rows {
        let res = sqlx::query(
            "INSERT OR IGNORE INTO fx_cache (pair, date, rate) VALUES ('USDJPY', ?, ?)"
        )
        .bind(date)
        .bind(rate)
        .execute(&mut *tx)
        .await?;
        inserted += res.rows_affected() as usize;
    }
    tx.commit().await?;
    Ok(inserted)
}

pub async fn get_usdjpy_on_or_before(
    pool: &SqlitePool,
    date: NaiveDate,
) -> Result<Option<f64>, sqlx::Error> {
    let row: Option<(f64,)> = sqlx::query_as(
        "SELECT rate FROM fx_cache WHERE pair = 'USDJPY' AND date <= ? ORDER BY date DESC LIMIT 1"
    )
    .bind(date)
    .fetch_optional(pool)
    .await?;
    Ok(row.map(|r| r.0))
}

File: src/db/queries/fx.ts
import { and, desc, eq, lte } from "drizzle-orm";
import { db, schema } from "@/db";

const { fxCache } = schema;
const PAIR = "USDJPY";

export function getLatestUsdJpy(): number | null {
  const row = db
    .select({ rate: fxCache.rate })
    .from(fxCache)
    .where(eq(fxCache.pair, PAIR))
    .orderBy(desc(fxCache.date))
    .limit(1)
    .get();
  return row?.rate ?? null;
}

export function bulkInsertUsdJpy(rows: [string, number][]): number {
  if (rows.length === 0) return 0;
  return db.transaction((tx) => {
    let inserted = 0;
    for (const [date, rate] of rows) {
      inserted += tx
        .insert(fxCache)
        .values({ pair: PAIR, date, rate })
        .onConflictDoNothing()
        .run().changes;
    }
    return inserted;
  });
}

export function getUsdJpyOnOrBefore(date: string): number | null {
  const row = db
    .select({ rate: fxCache.rate })
    .from(fxCache)
    .where(and(eq(fxCache.pair, PAIR), lte(fxCache.date, date)))
    .orderBy(desc(fxCache.date))
    .limit(1)
    .get();
  return row?.rate ?? null;
}

/** Upsert a single USDJPY rate (ON CONFLICT DO UPDATE) — used by daily/refresh. */
export function upsertUsdJpy(date: string, rate: number): void {
  db
    .insert(fxCache)
    .values({ pair: PAIR, date, rate })
    .onConflictDoUpdate({
      target: [fxCache.pair, fxCache.date],
      set: { rate },
    })
    .run();
}
